src/aind_metadata_upgrader/utils.py

```diff
@@ -0,0 +1,75 @@
+import logging
+
+from aind_data_schema.base import AindModel
+from aind_data_schema.core.procedures import (
+    Craniotomy,
+    FiberImplant,
+    Headframe,
+    IntraCerebellarVentricleInjection,
+    IntraCisternalMagnaInjection,
+    IntraperitonealInjection,
+    IontophoresisInjection,
+    NanojectInjection,
+    OphysProbe,
+    OtherSubjectProcedure,
+    Perfusion,
+    RetroOrbitalInjection,
+)
+from pydantic import ValidationError
+from pydantic.fields import PydanticUndefined
+
+
+def construct_new_model(model_inputs: dict, model_type: AindModel, allow_validation_errors=False):
+    """Validate a model, if it fails and validation error flag is on, construct a model"""
+
+    try:
+        return model_type.model_validate(model_inputs)
+    except ValidationError as e:
+        logging.error(f"Validation error in {type(model_type)}: {e}")
+        return model_type.model_construct(**model_inputs)
+
+
+def get_or_default(model: dict, model_type: AindModel, field_name: str, kwargs: dict = {}):
+    """Version of get_or_default that works with a dict instead of a model instance. If field is not explicitly set, will attempt to extract from a model."""
+
+    if kwargs.get(field_name) is not None:
+        return kwargs.get(field_name)
+    elif model.get(field_name, None) is not None:
+        return model.get(field_name)
+    else:
+        try:
+            attr_default = getattr(model_type.model_fields.get(field_name), "default")
+            if attr_default == PydanticUndefined:
+                return None
+            return attr_default
+        except AttributeError:
+            return None
+
+
+procedure_types_list = {
+    "Craniotomy": Craniotomy,
+    "Fiber implant": FiberImplant,
+    "Headframe": Headframe,
+    "Intra cerebellar ventricle injection": IntraCerebellarVentricleInjection,
+    "Intra cisternal magna injection": IntraCisternalMagnaInjection,
+    "Intraperitoneal injection": IntraperitonealInjection,
+    "Iontophoresis injection": IontophoresisInjection,
+    "Nanoject injection": NanojectInjection,
+    "Perfusion": Perfusion,
+    "Other subject procedure": OtherSubjectProcedure,
+    "Retro-orbital injection": RetroOrbitalInjection,
+    "ophys_probe": OphysProbe,
+}
+
+
+def drop_unused_fields(item, model_type):
```

Comment:
After looking at this a bit more, I think you should not need `drop_unused_fields` at all. The dictionaries you are creating as input to `construct_new_model` should be the correct shape - you should not need to drop anything.

Reply:
This is mostly only relevant for items where they are going to fail validation, but contain outdated fields, to make sure they don't populate into construction. If that doesn't seem like a worry, I can remove it. It's mostly deprecated and only being used in 1 or 2 areas.